### ISSUE TITLE
Psuedoinverse solves

### ISSUE THREAD

**Issue by shoyer** [google/jax#2802]
xref https://github.com/google/jax/pull/2794

If we're willing to define a "pseudo-inverse solve" (which as far as I can tell does not exist in NumPy or SciPy), we can speed-up gradients of pseudo-inverse solves by large factor (e.g., 10x) as the following micro-benchmark shows:
```python
import jax.numpy as jnp
import jax
from functools import partial
from jax import lax
import jax.numpy as jnp
import numpy as np

@jax.jit
def psolve(a, b):
  return jnp.linalg.pinv(a) @ b

@jax.jit
def psolve_custom(a, b):
  a_inv = jnp.linalg.pinv(a)
  a_t_inv = jnp.linalg.pinv(a.T)
  matvec = partial(jnp.matmul, a)
  solve = lambda matvec, x: a_inv @ x
  transpose_solve = lambda vecmat, x: a_t_inv @ x
  return lax.custom_linear_solve(matvec, b, solve, transpose_solve)

a = jnp.asarray(np.random.RandomState(0).randn(200, 200))
b = jnp.asarray(np.random.RandomState(1).randn(200))

jac_psolve = jax.jit(jax.jacobian(psolve))
jac_psolve_custom = jax.jit(jax.jacobian(psolve_custom))

%timeit jac_psolve(a, b).block_until_ready()  # 391 ms per loop
%timeit jac_psolve_custom(a, b).block_until_ready()  # 36 ms per loop
```

Should we go ahead and add a helper function for this somewhere? Maybe `jax.ops.linalg`?? We can add a loud WARNING to the docstring for `jnp.linalg.pinv`.

I suspect it simply doesn't exist in NumPy/SciPy because there isn't much to be gained for such a function if you're only worrying about the forward solve.

**Comment by mattjj** [comment]
Wow! I'm trying to grok this, in particular where the savings are coming from. It's not from reusing the SVD factorization of `a` again for the SVD of `a.T` (which I think we could do, right?), but something else...

**Comment by mattjj** [comment]
Okay, I think it's that as per Sec 2.3.1 of [the Mike Giles pdf](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf) I always return to for remembering this stuff, for `C = A^{-1} B` (like a solve primitive) we can write `Cdot = A^{-1}(Bdot - Adot C)`, whereas if we differentiate the composition of inverting and then multiplying we'd end up evaluating something that looks more like `Cdot = -(A^{-1} Adot C^{-1}) B + A^{-1} Bdot`. Does that sound right?

**Comment by shoyer** [comment]
Yes, that intuition looks about right to me, at least for inverses -- you turn matrix-matrix multiplications into matrix-vector multiplications.

The JVP rules for `svd` and `pinv` involve lots of dense matrix-matrix multiplication, which is slow. The performance gap is smaller on GPU, more like 95 ms vs. 28 ms. This makes sense because GPUs are faster for matmuls but not really faster for matrix decomposition.

**Comment by mattjj** [comment]
By the way, should the solves just form `U, s, V = svd(a)` and then multiply by `U @ (np.where(s > cutoff, np.divide(1, s), 0.) * V)` and `V.T @ (np.where(s > cutoff, np.divide(1, s), 0.) * U.T)` respecively?

**Comment by shoyer** [comment]
Yes, that looks about right. Definitely no need to do an `svd` twice.

**Comment by mattjj** [comment]
I tried it out and got got some issues, but maybe morally there's something to do there.

**Comment by shoyer** [comment]
I made a mistake in my benchmark script on GPUs earlier (I edited my earlier comment). It's only something like a 25% speed-up on GPUs, and only if we manage to avoid computing the SVD twice.

**Comment by shoyer** [comment]
Actually, I think I was just wrong here with my math. This worked on my test problem only because my random matrices happened to be invertible.

The essential idea of the trick for linear solves (as Matt notes in his comment) is just evaluating the gradient with operations in a different order, which allows for replacing matrix-matrix products with matrix-vector products and reusing more computation from the forward pass.

I think we could use essentially the same trick for pseudoinverse solves (and also reuse the SVD for calculating all the psuedoinverses we need in the gradient), but `custom_linear_solve` won't cut it.
